### ISSUE TITLE
TFLite in C++ causes Segmentation Fault for MobileNet

### ISSUE THREAD

**Issue by damjandakic93** [tensorflow/tensorflow#70802]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

binary

### TensorFlow version

2.13.1
2.15.1 behaves similarly
2.16.1 cannot even produce the tflite model (something similar to this issue: https://github.com/tensorflow/tensorflow/issues/65012)
tf-nightly cannot build libtensorflowlite.so (gcc: error: unrecognized command-line option '-mavx512fp16'; did you mean '-mavx512bf16'?)

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04.4 LTS

### Python version

3.10.12

### Current behavior?

I make a simple MobileNet model in TF, convert it to TFLite and attempt to run it in C++. It causes Segmentation Fault. Running TFLite Interpreter in Python works properly. I have created a minimal repro case:

```
import tensorflow as tf

def main():
    model = tf.keras.Sequential([
        tf.keras.Input(batch_shape=[1, 224, 224, 3]),
        tf.keras.applications.MobileNetV3Small(
            input_shape=[224, 224, 3],
            alpha=1.0,
            minimalistic=False,
            include_top=False,
            weights="imagenet",
            pooling="max",
            dropout_rate=0.2,
            classifier_activation=None,
            include_preprocessing=True,
        )
    ])

    converter = tf.lite.TFLiteConverter.from_keras_model(model)
    tflite_model = converter.convert()
    with open("converted_model.tflite", "wb") as f:
        f.write(tflite_model)


if __name__ == "__main__":
    main()
```

### Standalone code to reproduce the issue

```shell
#include <memory>
#include <iostream>
#include "tensorflow/lite/core/interpreter.h"
#include "tensorflow/lite/core/model_builder.h"
#include "tensorflow/lite/kernels/register.h"
#include "tensorflow/lite/core/c/common.h"
#include "tensorflow/lite/optional_debug_tools.h"
#include "flatbuffers/flatbuffers.h"

std::unique_ptr<tflite::Interpreter> Initialize()
{
    const std::string model_path = "converted_model.tflite";
    std::unique_ptr<tflite::FlatBufferModel> model =
        tflite::FlatBufferModel::BuildFromFile(model_path.c_str());

    tflite::ops::builtin::BuiltinOpResolver resolver;
    std::unique_ptr<tflite::Interpreter> model_interpreter;
    tflite::InterpreterBuilder(*model, resolver)(&model_interpreter);
    if (model_interpreter->AllocateTensors() != kTfLiteOk)
    {
        std::cerr << "Failed to allocate tensors." << std::endl;
    }

    return model_interpreter;
}

int main()
{
    std::unique_ptr<tflite::Interpreter> interpreter = Initialize();

    std::cout << "Acquiring input buffer" << std::endl;

    float *input_buff = interpreter->typed_input_tensor<float>(0);

    for (int i = 0; i < 224*224*3; i++)
    {
        input_buff[i] = 0.2;
    }

    std::cout << "Filled input buffer" << std::endl;

    interpreter->Invoke();

    std::cout << "Done with Invoke" << std::endl;

    float *output = interpreter->typed_output_tensor<float>(0);
    std::cout << "Output: " << output[0] << std::endl;

    return 0;
}
```

### Relevant log output

```shell
INFO: Initialized TensorFlow Lite runtime.
INFO: Applying 1 TensorFlow Lite delegate(s) lazily.
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
VERBOSE: Replacing 109 out of 110 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 2 partitions for the whole graph.
Note (XNNPACK): fuse Constant Pad Node #5 into Depthwise Convolution 2D Node #4
Note (XNNPACK): fuse Constant Pad Node #14 into Depthwise Convolution 2D Node #13
Note (XNNPACK): fuse Constant Pad Node #23 into Depthwise Convolution 2D Node #22
Note (XNNPACK): fuse Constant Pad Node #77 into Depthwise Convolution 2D Node #76
INFO: Successfully applied the default TensorFlow Lite delegate indexed at 0.
 *NOTE*: because a delegate has been applied, the precision of computations should be unchanged, but the exact output tensor values may have changed. If such output values are checked in your code, like in your tests etc., please consider increasing error tolerance for the check.
Acquiring input buffer
Filled input buffer
Segmentation fault (core dumped)
```

**Comment by damjandakic93** [comment]
Sure, though you have the code to generate it above.
Here it is attached.
[converted_model.tflite.zip](https://github.com/user-attachments/files/16264096/converted_model.tflite.zip)
Thanks!

**Comment by sawantkumar** [comment]
Hi @pkgoogle ,

I tried several times but i keep getting the below error where it says it cannot find the interpreter.h even though its present at that location, , can you please take a look?

```
CMake Generate step failed.  Build files cannot be regenerated correctly.
root@tflite-issue-replication:/home/sawantkumar/work/tflite_program/build# cd ..
root@tflite-issue-replication:/home/sawantkumar/work/tflite_program# nano CMakeLists.txt 
root@tflite-issue-replication:/home/sawantkumar/work/tflite_program# cd build/
root@tflite-issue-replication:/home/sawantkumar/work/tflite_program/build# cmake ..
-- Configuring done
-- Generating done
-- Build files have been written to: /home/sawantkumar/work/tflite_program/build
root@tflite-issue-replication:/home/sawantkumar/work/tflite_program/build# make
Scanning dependencies of target tflite_program
[ 50%] Building CXX object CMakeFiles/tflite_program.dir/model_runner.cpp.o
/home/sawantkumar/work/tflite_program/model_runner.cpp:3:10: fatal error: tensorflow/lite/core/interpreter.h: No such file or directory
    3 | #include "tensorflow/lite/core/interpreter.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/tflite_program.dir/build.make:63: CMakeFiles/tflite_program.dir/model_runner.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/tflite_program.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

**Comment by pkgoogle** [comment]
Hi @damjandakic93, are you willing to switch to a different workflow? Using [AI-Edge-Torch](https://github.com/google-ai-edge/ai-edge-torch) and our [minimal C++ example](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/examples/minimal), I was able to successfully run this model:

conversion:
```py
import ai_edge_torch
import torch
import torchvision


orig_model = torchvision.models.mobilenet_v3_small()
sample_input = (torch.randn(1, 3, 224, 224),)

edge_model = ai_edge_torch.convert(orig_model.eval(), sample_input)
edge_model.export("mobilenet_v3_small.tflite")
```

execution:
```sh
git clone https://github.com/tensorflow/tensorflow.git tensorflow_src
mkdir minimal_build
cd minimal_build
cmake ../tensorflow_src/tensorflow/lite/examples/minimal
cmake --build . -j
./minimal <path/to/mobilenet_v3_small.tflite> #wherever you saved it
```

example output:
```
./minimal xxxxxxxx/issues/tflite/70802/mobilenet_v3_small.tflite
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
=== Pre-invoke Interpreter State ===
Interpreter has 1 subgraphs.

-----------Subgraph-0 has 316 tensors and 132 nodes------------
1 Inputs: [0] -> 602112B (0.57MB)
1 Outputs: [222] -> 4000B (0.00MB)

Tensor  ID Name                      Type            AllocType          Size (Bytes/MB)    Shape      MemAddr-Offset  
Tensor   0 serving_default_args_0:0  kTfLiteFloat32  kTfLiteArenaRw     602112   / 0.57 [1,3,224,224] [0, 602112)
Tensor   1 arith.constant            kTfLiteFloat32  kTfLiteMmapRo      2359296  / 2.25 [1024,576] [7758680, 10117976)
Tensor   2 arith.constant1           kTfLiteFloat32  kTfLiteMmapRo      221184   / 0.21 [576,1,1,96] [7537476, 7758660)
Tensor   3 arith.constant2           kTfLiteFloat32  kTfLiteMmapRo      221184   / 0.21 [96,1,1,576] [7316280, 7537464)
Tensor   4 arith.constant3           kTfLiteFloat32  kTfLiteMmapRo      57600    / 0.05 [1,5,5,576] [7258668, 7316268)
Tensor   5 arith.constant4           kTfLiteFloat32  kTfLiteMmapRo      221184   / 0.21 [576,1,1,96] [7037472, 7258656)
Tensor   6 arith.constant5           kTfLiteFloat32  kTfLiteMmapRo      221184   / 0.21 [96,1,1,576] [6816276, 7037460)
Tensor   7 arith.constant6           kTfLiteFloat32  kTfLiteMmapRo      57600    / 0.05 [1,5,5,576] [6758664, 6816264)
Tensor   8 arith.constant7           kTfLiteFloat32  kTfLiteMmapRo      2304     / 0.00 [576] [6756348, 6758652)
Tensor   9 arith.constant8           kTfLiteFloat32  kTfLiteMmapRo      221184   / 0.21 [576,1,1,96] [6535152, 6756336)
Tensor  10 arith.constant9           kTfLiteFloat32  kTfLiteMmapRo      110592   / 0.11 [96,1,1,288] [6424548, 6535140)
Tensor  11 arith.constant10          kTfLiteFloat32  kTfLiteMmapRo      28800    / 0.03 [1,5,5,288] [6395736, 6424536)
...
...
...
Node 127 Operator Builtin Code  74 SUM (delegated by node 131)
  2 Input Tensors:[218,59] -> 0B (0.00MB)
  1 Output Tensors:[219] -> 0B (0.00MB)
  4 Temporary Tensors:[259-262] -> 0B (0.00MB)
Node 128 Operator Builtin Code   9 FULLY_CONNECTED (delegated by node 131)
  3 Input Tensors:[219,1,-1] -> 0B (0.00MB)
  1 Output Tensors:[220] -> 0B (0.00MB)
Node 129 Operator Builtin Code 117 HARD_SWISH (delegated by node 131)
  1 Input Tensors:[220] -> 0B (0.00MB)
  1 Output Tensors:[221] -> 0B (0.00MB)
Node 130 Operator Builtin Code   9 FULLY_CONNECTED (delegated by node 131)
  3 Input Tensors:[221,58,-1] -> 0B (0.00MB)
  1 Output Tensors:[222] -> 0B (0.00MB)
Node 131 Operator Custom Name TfLiteXNNPackDelegate 
  92 Input Tensors:[0-91] -> 10719000B (10.22MB)
  1 Output Tensors:[222] -> 4000B (0.00MB)

Execution plan as the list of 1 nodes invoked in-order: [131]
Among these nodes in the execution plan:
  Node 131 is a TfLiteXNNPackDelegate node (0x555e556d2f20), which has delegated 131 nodes: [0-130]
--------------Subgraph-0 dump has completed--------------

--------------Memory Arena Status Start--------------
Total memory usage: 606112 bytes (0.578 MB)
- Total arena memory usage: 606112 bytes (0.578 MB)
- Total dynamic memory usage: 0 bytes (0.000 MB)

Subgraph#0   Arena (Normal)         606112 (100.00%)
--------------Memory Arena Status End--------------

```

**Comment by damjandakic93** [comment]
@pkgoogle Wow, this seems interesting, thanks! I actually had a significant amount of work to switch from PyTorch I was working with to TensorFlow as I was unaware of this tool. The reason I had to switch was because ultimately I wish to run the model on a Coral device. Does this workflow support running models on Coral?

Also, my model is not a simple mobilenet like in this example but a custom model which relies on the mobilenet as a backbone (so mobilenet + some fully connected layers in several branches with regression outputs at the end), does this sound like something I'd be able to get running on Coral with AI-Edge-Torch?

**Comment by pkgoogle** [comment]
Hi @damjandakic93, ideally it should work with any model you can define in PyTorch, in practice this is probably not 100% true -- if it isn't, that team would love to hear from you :) so we can make the product even better. The produced tflite model is not particularly special/not-special so the produced model should run on Coral.

**Comment by damjandakic93** [comment]
Managed to convert the model (it seems v2.functional.rotate isn't supported but I've removed it for the sake of getting anything to work).
However, once I try to run it in tflite (C++, natively without TPU) it causes SegmentationFault on Invoke (same inference code as above, I've just commented-out the filling of the input buffer).
Now I cannot share with you this specific model (due to NDA). I can tell you that it contains MobileNet, slicing, concatenation, torch.nn.Linear and torch.nn.functional.silu which are all fairly simple operators.

Any hints on how to debug this futher? Thanks!

**Comment by damjandakic93** [comment]
Btw, custom matmul ("my_mat_mul.tflite") from the example conversion you linked above works properly so the issue is model-specific.

**Comment by damjandakic93** [comment]
Actually, let me rephrase my question.
Since I cannot give you the whole model, how do you suggest I proceed with the debug?
I an start from simple mobilenet and butcher my model until it starts working to see which op causes the segfault, or is there a better way? Thanks!

**Comment by pkgoogle** [comment]
@damjandakic93 What you are suggesting will work to isolate where the situation is happening though there is a risk that the system together is what causing the issue, but you will answer that as you try to isolate the issue. From there you will have a minimally reproducible model which will help you in the next step.

I'm guessing you have a program that is running the model in C++. You have to compile TF from source with debug symbols, so follow https://www.tensorflow.org/install/source and add this option when using Bazel: `--config=dbg`, install the version of TF you want to debug (I recommend starting with nightly as that represents the latest code, in case your issue is actually resolved already)

Choose your favorite debugger (usually this means gdb or lldb), use it with your program i.e. if you would execute your program like: `./your_program`, do `gdb ./your_program`

If you are unfamiliar with a debugger, well then it's time to start but I would look up a cheat sheet online and just start stepping through the code.

**Comment by damjandakic93** [comment]
I'll need some time (other priorities) before I dive into this debug (as it will take some time most likely).
I'll leave the issue as stale (and then closed) and will reopen it once I have an update.

**Comment by damjandakic93** [comment]
Hi @pkgoogle,

So, I came back back to this issue. I've relaxed my requirements to supporting Python as well (so no need for C++ anymore, it's too much of a hassle, I'll do it inter-process).
I start with a simple vanilla Mobilenet model:

`model = torchvision.models.get_model("mobilenet_v3_small", weights="DEFAULT")`

I convert it via aiedge:

```
model = ai_edge_torch.to_channel_last_io(model, args=[0])
_args = (
    torch.randn((1, 224, 224, 3), dtype=torch.float32),
)

edge_model = ai_edge_torch.convert(model, _args)
edge_model.export("edge_model.tflite")
```

When I try to run it through edgetpu_compiler via CLI I get the following:

> edgetpu_compiler edge_model.tflite 
> /bin/bash: /home/ddakic/anaconda3/envs/aiedge/lib/libtinfo.so.6: no version information available (required by /bin/bash)
> Edge TPU Compiler version 16.0.384591198
> ERROR: Didn't find op for builtin opcode 'MUL' version '7'. An older version of this builtin might be supported. Are you using an old TFLite binary with a newer model?
> 
> ERROR: Registration failed.
> 
> Invalid model: edge_model.tflite
> Model could not be parsed

Is there a way to control the opset version during the conversion (I guess that would be a reasonable path to start debugging)?

Thanks!

**Comment by pkgoogle** [comment]
Hi @damjandakic93, this looks like an edgetpu_compiler issue... is there a more relevant repo for that particular piece? Perhaps this? https://github.com/google-coral/libedgetpu/issues

**Comment by damjandakic93** [comment]
I'll try to debug it with them as well.
Thought this might also be the place to mention this as the same model implemented in TensorFlow and converted directly through TFLite converter works properly with the edgetpu_compiler. Let me try to work it out with them first then, in case you get some ideas on your side let me know please. Thanks!

**Comment by pkgoogle** [comment]
Hi @damjandakic93, then .. it may be an ai-edge-torch issue: https://github.com/google-ai-edge/ai-edge-torch/issues, I do think they can perhaps identify root cause better, if there's an issue w/ the converted .tflite file then route it to AI-Edge-Torch (i.e. it converted something improperly). But digging into why edgetpu_compiler is failing with it will be better handled by that repo for now.

**Comment by harshad2112** [comment]
I might be too late for this, but if anyone else wondering.

My friend was also getting seg fault in C++ code, and as we were debugging the issue we found out that the model that we are creating 

```
std::unique_ptr<tflite::FlatBufferModel> model = tflite::FlatBufferModel::BuildFromFile(model_path.c_str());
```

is unique pointer, and this gets passed to 

```
InterpreterBuilder(const FlatBufferModel& model, const OpResolver& op_resolver)
```

and as since interpreterBuilder takes model by reference, it does not take the ownership of the model, and as the Initialize() function finishes model also gets cleaned up (as it is a unique pointer) making the interpreter to point to some garbage value.

There can be several ways to resolve this, 
1. you can either release the model from unique pointer and assign it to raw pointer and then pass it to interpreterBuiler()
```
auto model_raw = model.release();  
tflite::InterpreterBuilder(*model_raw, resolver)(&model_interpreter);
```

2. return model as well from the initialize() function to main function
3. make a class with model and interpreter as private members. (Might be a overkill, but depends on your implementation).

I think tensorflow can change the builder class to take ownership of model also, so it does not get destroyed on function completion.
